Use goleak to capture goroutine leak

Ashmita152 opened the pull request:
Signed-off-by: Ashmita Bohara <ashmita.bohara152@gmail.com>

<!--
We appreciate your contribution to the Jaeger project! 👋🎉

Before creating a pull request, please make sure:
- Your PR is solving one problem
- You have read the guide for contributing
  - See https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING.md
- You signed all your commits (otherwise we won't be able to merge the PR)
  - See https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md#certificate-of-origin---sign-your-work
- You added unit tests for the new functionality
- You mention in the PR description which issue it is addressing, e.g. "Resolves #123"
-->

## Which problem is this PR solving?
This PR uses goleak (https://github.com/uber-go/goleak) to identify goroutine leak as part of unit tests.

## Short description of the changes
We are adding a `TestMain` function for each test package. Goleak provides a way to exclude top functions, hence we are adding `TolerantVerifyLeakMain` with those exceptions. If this is accepted change, I'll add changes for rest of the test packages too.

yurishkuro commented on pkg/testutils/testutils.go:
```diff
@@ -0,0 +1,46 @@
+func TolerantVerifyLeakMain(m *testing.M) {
+	goleak.VerifyTestMain(m,
+		// https://github.com/census-instrumentation/opencensus-go/blob/d7677d6af5953e0506ac4c08f349c62b917a443a/stats/view/worker.go#L34
+		goleak.IgnoreTopFunction("go.opencensus.io/stats/view.(*worker).start"),
```
where do we even use opencensus?

yurishkuro commented on pkg/testutils/testutils.go:
```diff
@@ -0,0 +1,46 @@
+
+// TolerantVerifyLeakMain verifies go leaks but excludes the go routines that are
+// launched as side effects of some of our dependencies.
+func TolerantVerifyLeakMain(m *testing.M) {
```
This is unfortunate. Seems to indicate that the tests are not closing the resources properly, or the dependencies are not exposing proper clean-up functions.

yurishkuro commented on pkg/testutils/testutils.go:
```diff
@@ -0,0 +1,46 @@
+		goleak.IgnoreTopFunction("k8s.io/klog/v2.(*loggingT).flushDaemon"),
+		goleak.IgnoreTopFunction("k8s.io/klog.(*loggingT).flushDaemon"),
+		// https://github.com/dgraph-io/badger/blob/v1.5.3/y/watermark.go#L54
+		goleak.IgnoreTopFunction("github.com/dgraph-io/badger/y.(*WaterMark).process"),
```
if feels like badger should really have proper clean-ups, maybe we're not invoking them?

yurishkuro commented on cmd/flags/flags_test.go:
```diff
@@ -21,8 +21,13 @@ import (
+	"go.uber.org/goleak"
 )
 
+func TestMain(m *testing.M) {
```
Manually adding these functions is not a scalable approach, next time someone will surely forget to add one.

I would prefer to have a CI check, similar to `make nocover`, that will fail if some packages don't have this function (assuming we want to go with checking these leaks).

yurishkuro reviewed (commented):
What is the motivation for using goleak? Does it really help identify issues that are not specific to the tests themselves? E.g. by adding these calls throughout, did you actually run into places where GRs were leaked (not counting external dependencies)?

Ashmita152 commented:
I am very new to golang but I saw goleak is adopted by multiple open-source golang projects like thanos. I just thought it will be good to catch such leaks in unittests. I have no experience with it. 
Possibly the issues with badger we noticed as part of this PR when using goleak with badgerexporter test is maybe related to the behaviour noticed in this https://github.com/jaegertracing/jaeger/issues/2459

yurishkuro commented:
How did you arrive at the exclusion list, by testing packages individually? I would prefer not to have any exclusions and just not enable goleaks for packages that leak, for now.

Ashmita152 commented:
Right, I added that `TestMain` at each package one by one and when it fails for some package like badgerexporter, it prints the top of the stack which I added to the exclusion list. Sure I will remove that.
